mattermost/mattermost-webapp
# MM-23367 Add test metadata to selectively run Cypress tests

#### Summary
__Changes included in this PR__
- add test metadata to group or skip tests like `// @prod @messaging @headless` --> To follow, will submit PR to docs
- moved spec files into its folder according to groups, mostly following tabs in the spreadsheet
- clean up spec headers
- assigned all 167 spec files as `@prod` and 50 as `@smoke` --> To follow, will sort out which specs should qualify as prod and smoke tests (to be decided by QA team).

__Test Metadata details__:
I initially designed grouping/filtering to be flexible up to `describe/context` and `it/test` block by using Mocha suite but there's some issue when running Cypress tests.  So, I ended up dong it on a higher level, which is per file basis.  Here are some of the details:
- Assign test metadata by adding comment per file
  - Add stage like `// Stage: @prod @smoke` which means a spec file is for production and smoke testing
  - Add group like `// Group: @accessibility`
  - Exclude test environment `Skip: @headless` which means a test can run in all environment except on headless mode
  - "Stage:", "Group:" and "Skip:" are written for guide only. The important keyword is "@" followed by a "word"
- Filtering spec files before passing into Cypress makes the reporting accurate since it doesn't need to mark a describe/test as skip or pending on runtime. It also prevents Cypress from unexpected error since it doesn't need to run hooks (e.g. before, after...) if test will eventually skipped.
- Easier to invert matches. If there are 7/10 assigned as `@prod`, inverting it by `node run_tests --stage='@prod' --invert` will run test to other 3/10 spec files. Those unassigned spec files can be from a newly submitted PR.

__How to run__:
1. Run with `node run_tests.js` - tests all spec files
2. Add `--stage` flag, like `node run_tests.js --stage='@prod'` - tests all files marked as `@prod`.  It can be of multiple values, comma-separated which will tests files if either of values were found.
<img width="829" alt="Screen Shot 2020-03-23 at 10 06 51 PM" src="https://user-images.githubusercontent.com/5334504/77325119-b6cf3480-6d52-11ea-9493-1ba68b08a6d4.png">
3. Add `--group` flag like `node run_tests.js --group='@accessibilty'`. Same logic as stage flag.
4. Add `--invert` to invert matches, like `node run_tests.js --group='@accessibilty' --invert` - test all files excluding `@accessibilty` group
5. Can combine all flags, like `node run_tests.js --stage='@prod @smoke ' --group='@messaging @accessibilty' --invert` - test all files excluding those that matches either stage values AND either group values.
6. Test can be skipped by environment: platform, browser or mode (headless/headed). If a test should run only on Chrome (headed) on any platform, then assigning with `@headless @firefox @electron` will evaluate whether the test can run on a given environment.
<img width="537" alt="Screen Shot 2020-03-23 at 10 07 13 PM" src="https://user-images.githubusercontent.com/5334504/77325133-bf276f80-6d52-11ea-895a-0db44777550c.png">

#### Ticket Link
JIRA ticket: https://mattermost.atlassian.net/browse/MM-23367
